#!/usr/bin/env Rscript 

library(DESeq2)
library(tibble)
library(reshape2)
library(RColorBrewer)
library(pheatmap)
library(dplyr)
library(ggplot2)
library(optparse)

option_list = list(
  make_option(
    c("--counts_file"),
    type = "character",
    default = NULL,
    help = "dataset file name",
    metavar = "character"
  ),
  make_option(
    c("--annot_file"),
    type = "character",
    default = NULL,
    help = "dataset file name",
    metavar = "character"
  ),
  make_option(
    c("--outdir"),
    type = "character",
    default = "output_dir",
    help = "output directory [default= %default]",
    metavar = "character"
  )
)
opt_parser = OptionParser(option_list = option_list)
opt = parse_args(opt_parser)

#if (is.null(opt$counts_file)) {
#  print_help(opt_parser) stop("At least one argument must be supplied (input file)\n", call. = FALSE)
#}

setwd(opt$outdir)
library(here)

# Generate a folder where the results/plots will be stored
# Make sure that the working directory is set with setwd()
data_plots <- paste(Sys.Date(), "data_plots", sep = "_")
dir.create(here(data_plots), showWarnings = TRUE)

### Data Wrangling
counts_data <- read.table(opt$counts_file, header = FALSE, sep = '\t')
colnames(counts_data) <- c("gene","counts","sample")

# This table was generated by extracting information for genes and gene name from column 9 in the .gff3 file
ecoli_annot <- read.table(opt$annot_file, header = TRUE, sep = '\t')

# Filtering counts >= 50 for each gene per sample per state 
counts_data <- counts_data %>% filter(grepl("gene:",gene)) %>% filter(counts >= 50) %>% distinct(.keep_all = TRUE)

counts_data$gene <- gsub("gene:","",counts_data$gene)

counts_data <- tidyr::spread(counts_data,sample,counts, fill = 0)

rownames(counts_data) <- counts_data$gene

counts_data$gene <- NULL

counts_distrib <- melt(counts_data)

ggplot(counts_distrib, aes(x = value, fill = variable))+ geom_density(aes(alpha = 0.1))+theme_bw()+xlim(0,10000)+labs(title = "Raw count Distribution",x="raw counts")

ggsave(here(data_plots,paste0("RawCount_distribution.pdf")), width = 15, height = 8)

### Generating Metadata file

genotype <- c("ecoli_dh10","ecoli_dh10","ecoli_dh10","ecoli_dh10")

condition <- c("state1","state1","state2","state2")

ecoli_dh10_metadata <- data.frame(genotype, condition)

rownames(ecoli_dh10_metadata) <- c("ecoli_state1_merged_rep1","ecoli_state1_merged_rep2","ecoli_state2_merged_rep1","ecoli_state2_merged_rep2")

ecoli_dh10_metadata

### DE Analysis - QC - Normalization

# Check to see if the rownames in or metadata file is same as colnames in the raw count data
all(rownames(ecoli_dh10_metadata) == colnames(counts_data))

# Create a DESeq2 object
dds_ecoli_dh10 <- DESeqDataSetFromMatrix(countData = counts_data, colData = ecoli_dh10_metadata, design = ~ condition)

# Estimate size factors used to normalize the counts per gene per sample
dds_ecoli_dh10 <- estimateSizeFactors(dds_ecoli_dh10)

# Extract the normalized counts from the DESeq2 object
norm_ecoli_dh10 <- counts(dds_ecoli_dh10, normalized = TRUE)

sizeFactors(dds_ecoli_dh10)

### DE Analysis - Unsupervised Clustering Analysis - Hierarchical Clustering & PCA
### Performed to identify outliers, other major sources of variation in the data and how similar are the samples to each other with regards to gene expression

### Log transformation of the normalized counts improves the visualization of the clustering. For RNA-Seq data DESeq2 uses a variant stabilizing clustering, which is a log transformation that moderates the variance across the mean
vsd_ecoli_dh10 <- vst(dds_ecoli_dh10, blind = TRUE)

### Generating a matrix of log2 normalized counts
vsd_mat <- assay(vsd_ecoli_dh10)

# Compute pairwise correlation values
vsd_cor <- cor(vsd_mat)

pheatmap(vsd_cor, annotation = select(ecoli_dh10_metadata, condition), filename = file.path(here(data_plots),"hierarchical_clustering.pdf"), width = 15, height = 8)

### PCA - finds the variation present in the dataset.
plotPCA(vsd_ecoli_dh10, intgroup = "condition")

ggsave(here(
  data_plots,
  paste0("PCA.pdf")
),
width = 6,
height = 6)

### Modeling raw counts for each gene
### Since we did not find any outliers, we proceed with running the Negative Binomial model on the DESEq2 object created earlier

# Run analysis
dds_ecoli_dh10 <- DESeq(dds_ecoli_dh10)

# How well did our data fit the model? The DESeq2 model uses the dispersion to assess the variability in expression when modeling the counts. This is done by plotting for dispersion estimates
# We expect the dispersion values to decrease with increasing mean in RNA Seq data. The plot above shows a cloud of data (black dots), which are the genes that don't entirely cluster around the maximum likelihood line (red line). This can be due to contamination or presence of variation in samples

pdf(here(data_plots,paste0("dispersion_estimate_plot.pdf")),width = 6,height = 6)
plotDispEsts(dds_ecoli_dh10)
dev.off()

# We can include a 1.25 log2FC threshold and test for significant genes. 
# Ideally, we need not apply the log2FC thresh, but when there are large # of significnat genes, 
# applying a threshold is useful.

ecoli_res <- results(dds_ecoli_dh10, contrast = c("condition","state1","state2"), alpha = 0.05, lfcThreshold = 0.32)
# Alpha denotes the significance level

# To improve the estimated Fold Change(FC), we can perform log2FC shrinkage - for genes with low amounts of information associated with them. Shrinkage uses information from all genes to generate more likely, lower log2 FC
ecoli_res <- lfcShrink(dds_ecoli_dh10, contrast = c("condition","state1","state2"), res = ecoli_res)

summary(ecoli_res)

### Obtaining the gene name information for the DE results

#Join the annotation table with the results
ecoli_res_all <- data.frame(ecoli_res) %>% rownames_to_column(.,var = "gene")

ecoli_annot$gene <- as.character(ecoli_annot$gene)
ecoli_res_all$gene <- as.character(ecoli_res_all$gene)

ecoli_res_all <- left_join(ecoli_res_all, ecoli_annot, by = "gene")

# Subset the significant genes
sig_ecoli_dh10 <- subset(ecoli_res_all, padj < 0.05) %>% filter(!is.na(name)) %>% arrange(padj)

#List of Significant genes
write.csv(sig_ecoli_dh10, file = "Significant_genes_between_the_2_states.csv", row.names = FALSE)

### Visualization of the results

# Expression Heatmap - plotting the normalized count values instead of the sample correlation values

# Subset the normalized counts of the significant genes
sig_norm_counts_ecoli_dh10 <- norm_ecoli_dh10[sig_ecoli_dh10$gene,]

write.csv(sig_norm_counts_ecoli_dh10, file = "Significant_genes_between_the_2_states_with_norm_couns.csv", row.names = TRUE)

heat_colors <- brewer.pal(6, "YlOrRd")

pheatmap(sig_norm_counts_ecoli_dh10,color = heat_colors,cluster_rows = T,show_rownames = F,annotation = select(ecoli_dh10_metadata, condition),scale = "row", filename = file.path(here(data_plots),"significant_genes_with_normalized_counts_expression_heatmap.pdf"), width = 15, height = 8)

# Volcano plot

# Obtain logical vector regarding whether padj values are less than 0.05 
ecoli_res_all <- ecoli_res_all %>% mutate(threshold = padj < 0.05)

ggplot(ecoli_res_all) + geom_point(aes(x = log2FoldChange, y = -log10(padj),color = threshold)) + xlab("log2 fold change") + ylab("-log10 adjusted p-value") + theme(legend.position = "none",plot.title = element_text(size = rel(1.5), hjust = 0.5), axis.title = element_text(size = rel(1.25)))+theme_bw()

ggsave(here(data_plots,"Volcano_plot_significant_genes.pdf"),width = 15, height = 8)

# Expression plot of the Top 20 significant genes

top_20 <- data.frame(sig_norm_counts_ecoli_dh10[1:20,]) %>% rownames_to_column(var = "gene")

top_20 <- tidyr::gather(top_20,key = "samplename",value = "normalized_counts",2:5)

# Join with the metadata
top_20 <- left_join(top_20,rownames_to_column(ecoli_dh10_metadata, var = "samplename"), by = "samplename")

ggplot(top_20) + geom_point(aes(x = gene, y = normalized_counts, color = condition)) + scale_y_log10() + xlab("Genes") + ylab("Normalized Counts") + ggtitle("Top 20 Significant DE Genes") + theme_bw() +  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + theme(plot.title = element_text(hjust = 0.5))

ggsave(here(data_plots,"Top20_significant_genes.pdf"),width = 15, height = 8)
