#!/usr/bin/env Rscript 

library(DESeq2)
library(tibble)
library(reshape2)
library(RColorBrewer)
library(pheatmap)
library(dplyr)
library(ggplot2)
library(optparse)   

option_list = list(
  make_option(
    c("--counts_file"),
    type = "character",
    default = NULL,
    help = "dataset file name",
    metavar = "character"
  ),
  make_option(
    c("--annot_file"),
    type = "character",
    default = NULL,
    help = "dataset file name",
    metavar = "character"
  ),
  make_option(
    c("--wkdir"),
    type = "character",
    default = "working_dir",
    help = "working directory [default= %default]",
    metavar = "character"
  )
)
opt_parser = OptionParser(option_list = option_list)
opt = parse_args(opt_parser)

if (is.null(opt$counts_file)) {
  print_help(opt_parser) stop("At least one argument must be supplied (input file).n", call. = FALSE)
}

setwd(opt$wkdir)

counts_data <- read.table(opt$counts_file, header = FALSE, sep = '\t')
colnames(counts_data) <- c("gene","counts","sample")

### This table was generated by extracting information for genes and gene name from column 9 in the .gff3 file
ecoli_annot <- read.table(opt$annot_file, header = TRUE, sep = '\t')

counts_data <- counts_data %>% filter(grepl("gene:",gene)) %>% filter(counts >= 50) %>% distinct(.keep_all = TRUE)

counts_data$gene <- gsub("gene:","",counts_data$gene)

counts_data <- tidyr::spread(counts_data,sample,counts, fill = 0)

rownames(counts_data) <- counts_data$gene

counts_data$gene <- NULL

counts_distrib <- melt(counts_data)

ggplot(counts_distrib, aes(x = value, fill = variable))+ geom_density(aes(alpha = 0.1))+theme_bw()+xlim(0,10000)+labs(title = "Raw count Distribution",x="raw counts")

ggsave("RawCount_distribution.pdf", width = 15, height = 8)

genotype <- c("ecoli_dh10","ecoli_dh10","ecoli_dh10","ecoli_dh10")

condition <- c("state1","state1","state2","state2")

ecoli_dh10_metadata <- data.frame(genotype, condition)

rownames(ecoli_dh10_metadata) <- c("ecoli_state1_merged_rep1","ecoli_state1_merged_rep2","ecoli_state2_merged_rep1","ecoli_state2_merged_rep2")


ecoli_dh10_metadata

# Check to see if the rownames in or metadata file is same as colnames in the raw count data
all(rownames(ecoli_dh10_metadata) == colnames(counts_data))

# Create a DESeq2 object
dds_ecoli_dh10 <- DESeqDataSetFromMatrix(countData = counts_data, colData = ecoli_dh10_metadata, design = ~ condition)

# Estimate size factors used to normalize the counts per gene per sample
dds_ecoli_dh10 <- estimateSizeFactors(dds_ecoli_dh10)

# Extract the normalized counts from the DESeq2 object
norm_ecoli_dh10 <- counts(dds_ecoli_dh10, normalized = TRUE)

sizeFactors(dds_ecoli_dh10)

vsd_ecoli_dh10 <- vst(dds_ecoli_dh10, blind = TRUE)

### Generating a matrix of log2 normalized counts
vsd_mat <- assay(vsd_ecoli_dh10)

# Compute pairwise correlation values
vsd_cor <- cor(vsd_mat)

v <- pheatmap(vsd_cor, annotation = select(ecoli_dh10_metadata, condition))

pheatmap(v, cellwidth = 15, cellheight = 8, filename = "hierarchical_clustering.pdf")

### PCA - finds the variation present in the dataset.
pdf(file = "PCA.pdf",   # The directory you want to save the file in
    width = 6, # The width of the plot in inches
    height = 6)
plotPCA(vsd_ecoli_dh10, intgroup = "condition")
dev.off()

# Run analysis
dds_ecoli_dh10 <- DESeq(dds_ecoli_dh10)

pdf(file = "dispersion_plot.pdf",
    width = 6,
    height = 6)
plotDispEsts(dds_ecoli_dh10)
dev.off()

# We can include a 1.25 log2FC threshold and test for significant genes. 
# Ideally, we need not apply the log2FC thresh, but when there are large # of significnat genes, 
# applying a threshold is useful.

ecoli_res <- results(dds_ecoli_dh10, contrast = c("condition","state1","state2"), alpha = 0.05, lfcThreshold = 0.32)
# Alpha denotes the significance level

ecoli_res <- lfcShrink(dds_ecoli_dh10, contrast = c("condition","state1","state2"), res = ecoli_res)

summary(ecoli_res)

#Join the annotation table with the results
ecoli_res_all <- data.frame(ecoli_res) %>% rownames_to_column(.,var = "gene")

ecoli_annot$gene <- as.character(ecoli_annot$gene)
ecoli_res_all$gene <- as.character(ecoli_res_all$gene)

ecoli_res_all <- left_join(ecoli_res_all, ecoli_annot, by = "gene")

# Subset the significant genes
sig_ecoli_dh10 <- subset(ecoli_res_all, padj < 0.05) %>% filter(!is.na(name)) %>% arrange(padj)

#List of Significant genes
write.csv(sig_ecoli_dh10, file = "Significant_genes_between_the_2_states.csv", row.names = FALSE)

# Subset the normalized counts of the significant genes
sig_norm_counts_ecoli_dh10 <- norm_ecoli_dh10[sig_ecoli_dh10$gene,]

heat_colors <- brewer.pal(6, "YlOrRd")

s <- pheatmap(sig_norm_counts_ecoli_dh10,color = heat_colors,cluster_rows = T,show_rownames = F,annotation = select(ecoli_dh10_metadata, condition),scale = "row")

pheatmap(s, cellwidth = 15, cellheight = 8, filename = "significant_genes_with_normalized_counts_expression_heatmap.pdf")

# Obtain logical vector regarding whether padj values are less than 0.05 
ecoli_res_all <- ecoli_res_all %>% mutate(threshold = padj < 0.05)

ggplot(ecoli_res_all) + geom_point(aes(x = log2FoldChange, y = -log10(padj),color = threshold)) + xlab("log2 fold change") + ylab("-log10 adjusted p-value") + theme(legend.position = "none",plot.title = element_text(size = rel(1.5), hjust = 0.5), axis.title = element_text(size = rel(1.25)))+theme_bw()

ggsave("Volcano_plot_significant_genes.pdf",width = 15, height = 8)

top_20 <- data.frame(sig_norm_counts_ecoli_dh10[1:20,]) %>% rownames_to_column(var = "gene")

top_20 <- tidyr::gather(top_20,key = "samplename",value = "normalized_counts",2:5)

# Join with the metadata
top_20 <- left_join(top_20,rownames_to_column(ecoli_dh10_metadata, var = "samplename"), by = "samplename")

ggplot(top_20) + geom_point(aes(x = gene, y = normalized_counts, color = condition)) + scale_y_log10() + xlab("Genes") + ylab("Normalized Counts") + ggtitle("Top 20 Significant DE Genes") + theme_bw() +  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + theme(plot.title = element_text(hjust = 0.5))

ggsave("Top20_significant_genes.pdf",width = 15, height = 8)
