# RNA_seq_DE_analysis_pipeline
This pipeline is used to run the RNA-Seq workflow for performing Differential Expression Analysis

**Note** Prior to running the workflow. There are a series of QC steps performed to eventually obtain the readcount matrix of genes for every sample.

### RNA Seq QC pipeline used
1. The split replicate fastq files were first merged so that we obtain 2 replicates per state.
2. Before proceeding with running the shell script. Place the `truseq_adapter.fasta`, `phix.fasta`, `plasmid.fasta`, `ecoli_dh10b_annotation.gff3`, `ecoli_dh10b_annotation.gtf` and the `ecoli_DH10B_ref_genome_NCBI_sequence.fasta` files in a location which can be accessed by the `QC_alignment_and_counts_pipeline.sh` file. it is preferred to keep all of them in the base directory location where all the new folders and files will be generated.
**Make sure that you update the QC_alignment_and_counts_pipeline.sh file to reflect the updated locations for all the .fasta files**
3. The `QC_alignment_and_counts_pipeline.sh` script file does the following things - 
* Removes TruSeq adapter reads, phix reads (a regular illumina spike-in) and plasmid sequences using the bbduk.sh script from the bbMap tools package
* FastQC stats is generated. 
The first 3 parts of the script generates 3 folders and 1 file `qc_reads`, `phix_qc_reads`, `plasmid_removed_reads` and `multiqc_report.html`. I generated a multiqc report of all the FastQC stats for every .fastq file in 1 report called `multiqc_report.html`. For further downstream processing, we are only interested in files present in `plasmid_removed_reads` folder.
* QC-ed fastq files present in the `plasmid_removed_reads` are further used to align to the ecoli reference genome by creating a genome index file and then proceeding with alignment using a splice-aware aligner called STAR, in the Part 4 section of the `QC_alignment_and_counts_pipeline.sh` script.
**Make sure that the ecoli reference fasta file location is properly mentioned while creating a genome index file**
* Once aligned to the reference genome, `htseq-counts` was used to count the genes mapped to the genome in Part 5 of the script.
* The sample specific count files are then merged into a count matrix file to feed through the DESeq2 RNA-Seq analysis pipeline in Part 6 of the script.
* You can simply run the script by typing - 
```
./QC_alignment_and_counts_pipeline.sh <dir_name>

where <dir_name> has all the merged  replicate fastq.gz sample files are located
```

* `QC_alignment_and_counts_pipeline.sh` is run on a linux based OS. All the required scripts and files to run the RNA-Seq QC pipeline can be found in the folder
* The merged count matrix file generated in Part 6 of the script is further used to run the RNA Seq DE analysis workflow to obtain the differential genes between the two states.

### Running the differential expression analysis workflow
1. This step can be run on a server with latest R server installed. 
**Note:** Make sure you have the latest version of R server installed.
2. Once you have confirmed the installation, type `R` in your command prompt (Windows/Linux) which will take you to the R environment.
3. Install all the necessary packages needed to run the R DESeq2 workflow by typing the following - 
```
install.packages(c("here","tibble","reshape2","RColorBrewer","pheatmap","dplyr","ggplot2","optparse"))
```
**To install DESeq2 you can refer the link - [DESeq2 installation](http://bioconductor.org/packages/release/bioc/html/DESeq2.html)**

4. Once all the packages are installed you are ready to run the workflow! 
**NOTE: Make sure the merged counts matrix file, the gene_annotation file and the .R script exist in the same folder**

5. Go to your command prompt and type - 
```
Rscript --vanilla Data_Challenge_Asimov.R --help
```
This will display the help section mentioning the command line argurments to use to run the workflow.

6. To run the workflow, simply type - 
```
Rscript --vanilla Data_Challenge_Asimov.R --counts_file <merged_counts.txt> --annot_file <geneIDs_name.txt> --outdir <the folder where all the 3 files exist> (as mentioned in the **NOTE** in step #4).
```

7. when you hit enter, the script runs and generates 1 folder and 2 output files. The folder consists of all the plots generated by the R workflow. The 2 files are the significant gene files with Log2 fold change and p.adjusted value information. The other file consists of the normalized counts for the significnat genes.
